lib/src/executor/offchain_runtime_host.rs

```diff
@@ -0,0 +1,536 @@
+            host::HostVm::OffchainSubmitTransaction(req) => {
+                let value = value.as_ref();
+                self.inner.vm = req.resume((value, value.len()));
+            }
+            // We only create a `OffchainSubmitTransaction` if the state is one of the above.
+            _ => unreachable!(),
+        };
+
+        self.inner.run()
+    }
+}
+
+/// Implementation detail of the execution. Shared by all the variants of [`OffchainRuntimeHostVm`]
+/// other than [`OffchainRuntimeHostVm::Finished`].
+struct Inner {
+    /// Virtual machine running the call.
+    vm: host::HostVm,
+    /// The changes made during execution.
+    pub overlay_storage: hashbrown::HashMap<Vec<u8>, Option<Vec<u8>>, fnv::FnvBuildHasher>,
+    /// Pending changes to the off-chain storage that this execution performs.
+    offchain_storage_changes: hashbrown::HashMap<Vec<u8>, Option<Vec<u8>>, fnv::FnvBuildHasher>,
+    /// Concatenation of all the log messages generated by the runtime.
+    logs: String,
+    /// Value provided by [`Config::max_log_level`].
+    max_log_level: u32,
+}
+
+impl Inner {
+    /// Continues the execution.
+    fn run(mut self) -> OffchainRuntimeHostVm {
+        loop {
+            match self.vm {
+                host::HostVm::ReadyToRun(r) => self.vm = r.run(),
+
+                host::HostVm::Error { error, prototype } => {
+                    return OffchainRuntimeHostVm::Finished(Err(Error {
+                        detail: ErrorDetail::WasmVm {
+                            error,
+                            logs: self.logs,
+                        },
+                        prototype,
+                    }));
+                }
+
+                host::HostVm::Finished(finished) => {
+                    return OffchainRuntimeHostVm::Finished(Ok(Success {
+                        virtual_machine: SuccessVirtualMachine(finished),
+                        offchain_storage_changes: self.offchain_storage_changes,
+                        logs: self.logs,
+                    }));
+                }
+
+                host::HostVm::ExternalStorageGet(req) => {
+                    let key = req.key().as_ref().to_vec();
+
+                    match self.overlay_storage.get(&key) {
+                        Some(value) => {
+                            self.vm = req.resume_full_value(value.as_ref().map(|v| &v[..]));
+                        }
+                        None => {
+                            self.vm = req.into();
+                            return OffchainRuntimeHostVm::StorageGet(StorageGet { inner: self });
+                        }
+                    }
+                }
+
+                host::HostVm::ExternalStorageSet(req) => {
+                    let key = req.key().as_ref().to_vec();
+                    let value = req.value().map(|v| v.as_ref().to_vec());
+                    self.overlay_storage.insert(key, value);
+                    self.vm = req.resume()
+                }
```

Comment:
I'm pretty sure the offchain worker isn't supposed to modify the storage during the execution, no?
Does this block of code ever get reached?

Reply:
yes it does. It will initialize the block and set current hash. technically it can clear storage as welll but should it be supported, I don't know

Reply:
https://github.com/paritytech/substrate/blob/6bae274f42520c38499e0091ebc50b68cff00c4e/frame/executive/src/lib.rs#L660-L677

Reply:
After some chatting, indeed offchain workers can do anything, and we just have to discard the overlay changes at the end.
(I've also opened https://github.com/smol-dot/smoldot/issues/828)

Sorry for the extra work, but please revert https://github.com/smol-dot/smoldot/pull/788/commits/87a2c2e2f2542c84ca3fc021a0c0d82abb0b6264 then :pray: 

I think it makes sense for the offchain-worker-related stuff to be in `runtime_host` as well, with the intention that `runtime_host` becomes the unique way to call a runtime.

However, instead of having one variant per offchain-worker-related function, I think the offchain-worker-related functions should be grouped in another enum.

In other words, not like this:

```rust
enum RuntimeHostVm {
    StorageGet,
    NextKey,
    ...
    OffchainGet,
    OffchainSeed,
    ...
}
```

But like that:

```rust
enum RuntimeHostVm {
    StorageGet,
    NextKey,
    ...
    Offchain(OffchainContext),
    ...
}

enum OffchainContext {
    Get,
    Seed,
    SubmitTransaction,
    ...
}
```

This makes it easier for the API user of `runtime_host` to error if an offchain worker function is called by the runtime.